### DISCUSSION TITLE
storybook 8 - react-dom-shim - Support for React 17

### DISCUSSION THREAD

**Discussion by nchery** [storybookjs/storybook#26826] [Help]
### Summary

I can't tell yet if it has to be considered as an issue so i ask a question.

I'm running into an issue where webpack compilation throwing an error `Module not found: Error: Can't resolve 'react-dom/client' in /node_modules/.pnpm/@storybook+react-dom-shim@8.0.6_react-dom@17.0.2_react@17.0.2/node_modules/@storybook/react-dom-shim/dist/react-18.mjs 2:0-40`

I'm on React 17 and react-dom 17 at the root of my project. If i understood correctly the code from react-dom-shim [(Repo)](https://github.com/storybookjs/storybook/blob/next/code/lib/react-dom-shim/src/preset.ts), and the comment line 5,

> /**
> * Get react-dom version from the resolvedReact preset, which points to either
> * a root react-dom dependency or the react-dom dependency shipped with addon-docs
>*/
 
on React 17 i should fall back to the file [(react-16.tsx)](https://github.com/storybookjs/storybook/blob/next/code/lib/react-dom-shim/src/react-16.tsx) which points to the former API 'import ReactDOM from 'react-dom';

However it is not the case, it points to react-18.mjs which calls the latest API of react-dom making the webpack compilation fails 

Any suggestions to where to look at to get this solved ?

### Additional information

The project is to update to Storybook 8 and the compilation error is thrown when starting cypress which is compiling some storybook stories. All dependencies were moved to 8.0.6. I'm webpack 5

``` =json
{
  "@storybook/addon-actions": "^8.0.6",
    "@storybook/addon-docs": "^8.0.6",
    "@storybook/addon-essentials": "^8.0.6",
    "@storybook/addon-links": "^8.0.6",
    "@storybook/addon-viewport": "^8.0.6",
    "@storybook/preview-api": "^8.0.6",
    "@storybook/nextjs": "^8.0.6",
    "@storybook/react": "^8.0.6",
}
```

**Comment by maasencioh** [comment]
@tmeasday as contributor of that function, I don't know if you could help us?

**Reply by tmeasday** [reply]
Hi @maasencioh -- I'm not sure about the `resolvedReact` part, that happened after I wrote this code. You could edit `node_modules` or use a debugger and figure out what the value of it is, or `require.resolve('react-dom/package.json')` returns inside that package?

Worst case, you could try using the `legacyRootApi` option to your framework to force React 17? https://storybook.js.org/docs/api/main-config/main-config-framework

**Comment by ej612** [comment]
I'm trying to update to Storybook 9 and am having similar errors that suggest something's expecting React 18, but I'm still stuck on 17:
```
ERROR in ../../node_modules/@storybook/addon-docs/dist/blocks.mjs 164:61504-61509
export 'useId' (imported as 'useId') was not found in 'react'

ERROR in ../../node_modules/@storybook/react-dom-shim/dist/react-18.mjs 2:0-45
Module not found: Error: Can't resolve 'react-dom/client' in 'C:\Users\xyz\node_modules\@storybook\react-dom-shim\dist'

ERROR in ../../node_modules/@storybook/react/dist/chunk-CQ7IB4U6.mjs 7:40453-40483
Module not found: Error: Can't resolve 'react-dom/test-utils' in 'C:\Users\xyz\node_modules\@storybook\react\dist'
Did you mean 'test-utils.js'?
BREAKING CHANGE: The request 'react-dom/test-utils' failed to resolve only because it was resolved as fully specified
(probably because the origin is strict EcmaScript Module, e. g. a module with javascript mimetype, a '*.mjs' file, or a '*.js' file where the package.json contains '"type": "module"').
The extension in the request is mandatory for it to be fully specified.
Add the extension to the request.

ERROR in ../../node_modules/storybook/dist/theming/index.js 1316:25-46
export 'useInsertionEffect' (imported as 'Pe') was not found in 'react'

ERROR in ../../node_modules/storybook/dist/theming/index.js 1316:49-70
export 'useInsertionEffect' (imported as 'Pe') was not found in 'react'
```

**Reply by ohaicristina** [reply]
I'm also experiencing this - specifically with the `useInsertionEffect` and `useId` - we're stuck on React 17 as well and according to the [Compatibility tracker](https://github.com/storybookjs/storybook/issues/23279), it should be compatible with as low of a version as 16
